### ISSUE TITLE
electric-pair and smartparens break copilot-completion-map

### ISSUE THREAD

**Issue by LemonBreezes** [zerolfx/copilot.el#86]
Hi. I have been having an issue where when I type '(' and the closing ')' is automatically inserted after point, the `copilot-completion-map` no longer becomes active. However, this issue does not seem to always occur with `electric-pair`, whereas with `smartparens`, it occurs much more reliably and goes away when I remove the `smartparens` `post-self-insert-hook`.

**Comment by JP-Ellis** [comment]
I can confirm that I have the same issue.

To help with some debugging:

- Completions work fine if there are no trailing closing parenthesis (both with and without `smartparens-mode`)
  ```rust
  impl error::Error for StepError {
      fn <TAB>
  }
  ```

- Completion fails if I have just inserted an opening parenthesis:
  ```rust
  impl error::Error for StepError {
      fn source(<TAB>
  }
  ```
    Instead of accepting the completion, `<TAB>` ends up inserting additional closing parenthesis with the result becoming:
  ```rust
  impl error::Error for StepError {
      fn source()))
  }
  ```

  Note that with the above, it only works when I have a 'pending' closing parenthesis (not sure how `smartparens-mode` calls that internally). So for example, if the code was already `fn source()` and I place my cursor in the middle of `()` and try and get a code completion, I am able to access the completion fine.

**Comment by ryankhart** [comment]
I am so glad I stumbled upon this open issue!

This was the source of my troubles with Copilot not working while inside parentheses (and other pairs of symbols that `smartparens-mode` helped automate like {}, [], '', "", and ``.

I'm using Doom Emacs, and I have below, a first draft suggestion for revising the Doom Emacs example config in the README for this repo.
```Elisp
;; accept completion from copilot and fallback to company
(use-package! copilot
  :hook (prog-mode . copilot-mode)
  :bind (("C-TAB" . 'copilot-accept-completion-by-word)
         ("C-<tab>" . 'copilot-accept-completion-by-word)
         :map copilot-completion-map
         ("<tab>" . 'copilot-accept-completion)
         ("TAB" . 'copilot-accept-completion))
  :config
  ;; Upon entering copilot mode, disable smartparens-mode as it is incompatible
  (add-hook 'copilot-mode-hook #'turn-off-smartparens-mode))
```
I don't have electric-pair installed, and I haven't looked up what the function call would be to turn that off, but that could go after the `:config` line as well.

I'd feel inclined to submit a pull request to update the README to include changes like these, but I'm such a novice at Emacs configs in general that I feel like I'd want a second opinion if flat out disabling `smartparens-mode` during `copilot-mode` is the best idea or not. I also don't know what would happen, with the config in my code block above, if you would leave `copilot-mode`. Would it automatically revert any changes that were made on the `'copilot-mode-hook`? Or would I have to write a separate hook of some sort to re-enable `smartparens-mode` after leaving `copilot-mode`?

**Comment by JP-Ellis** [comment]
I doubt that disabling smartparens-mode entirely is the best way forward. I have created an issue over in the smartparens repo to see if they have suggestions: Fuco1/smartparens#1145

**Comment by zerolfx** [comment]
@JP-Ellis @LemonBreezes I changed the priority of copilot's overlay. Could you verify the fix?

**Comment by ryankhart** [comment]
> @JP-Ellis @LemonBreezes I changed the priority of copilot's overlay. Could you verify the fix?

I believe it is fixed for me now with smartparens. I've removed the special mapping for `sp-pair-overlay-keymap` in my config.el file, restarted Emacs, and did my usual tests for copilot/smarparens compatibility by typing `(ma` and then <tab>, and it worked. It accepted the copilot suggestion rather than inserting an indentation.

Disclaimer: I'm still learning how package management works in Doom Emacs, so I'm not 100% which commit that I currently have installed, but I don't have it pinned and I've ran `doom sync` so I assume I have the latest commit.
